libs/libzauth/libzauth-c/src/lib.rs

```diff
@@ -284,8 +312,197 @@ impl From<TokenVerification> for ZauthTokenVerification {
+            uid: ptr::null(),
+            status: OAuthResultStatus::Utf8Error,
+        }
+    }
+}
+
+impl From<NulError> for OAuthResult {
+    fn from(_: NulError) -> Self {
+        OAuthResult {
+            uid: ptr::null(),
+            status: OAuthResultStatus::Panic,
+        }
+    }
+}
+
+impl From<OauthError> for OAuthResult {
+    fn from(e: OauthError) -> Self {
+        OAuthResult {
+            uid: ptr::null(),
+            status: e.into(),
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn oauth_key_open(
+    f: *const u8,
+    n: size_t,
+    k: *mut *mut OAuthJwk,
+) -> OAuthResultStatus {
+    if f.is_null() {
+        return OAuthResultStatus::NullArg;
+    }
+    catch_unwind_with(
+        || {
+            let bytes = unsafe { slice::from_raw_parts(f, n) };
+            let path = try_unwrap!(str::from_utf8(bytes));
+            let mut rdr = BufReader::new(try_unwrap!(File::open(&Path::new(path))));
+            let mut txt = String::new();
+            try_unwrap!(rdr.read_to_string(&mut txt));
+            unsafe {
+                *k = Box::into_raw(Box::new(OAuthJwk(txt)));
+            }
+            OAuthResultStatus::Ok
+        },
+        OAuthResultStatus::Panic,
+    )
+}
+
+#[no_mangle]
+pub extern "C" fn oauth_key_delete(a: *mut OAuthJwk) {
+    catch_unwind_with(
+        || {
+            unsafe {
+                drop(Box::from_raw(a));
+            }
+            OAuthResultStatus::Ok
+        },
+        OAuthResultStatus::Panic,
+    );
+}
+
+#[no_mangle]
+pub extern "C" fn oauth_verify_token(
+    jwk: &OAuthJwk,
+    token: *const u8,
+    token_len: size_t,
+    scope: *const u8,
+    scope_len: size_t,
+    method: *const u8,
+    method_len: size_t,
+) -> OAuthResult {
+    match panic::catch_unwind(|| {
+        if token.is_null() {
+            return OAuthResult {
+                uid: ptr::null(),
+                status: OAuthResultStatus::NullArg,
+            };
+        }
+        if scope.is_null() {
+            return OAuthResult {
+                uid: ptr::null(),
+                status: OAuthResultStatus::NullArg,
+            };
+        }
+        if method.is_null() {
+            return OAuthResult {
+                uid: ptr::null(),
+                status: OAuthResultStatus::NullArg,
+            };
+        }        
+        let bytes = unsafe { slice::from_raw_parts(token, token_len) };
+        let token = try_unwrap!(str::from_utf8(bytes));
+        let bytes = unsafe { slice::from_raw_parts(scope, scope_len) };
+        let scope = try_unwrap!(str::from_utf8(bytes));
+        let bytes = unsafe { slice::from_raw_parts(method, method_len) };
+        let method = str::from_utf8(bytes).unwrap();
+        let subject = try_unwrap!(verify_oauth_token(&jwk.0, token, scope, method));
+        let c_str = try_unwrap!(CString::new(subject));
+        OAuthResult {
+            uid: c_str.into_raw(),
+            status: OAuthResultStatus::Ok,
+        }
+    }) {
```

Comment:
Can't you use `catch_unwind_with` here too?

Reply:
Yes!